### ISSUE TITLE
Add ability to test rule schemas

### ISSUE THREAD

**Issue by bmish** [eslint/eslint#13434]
**The version of ESLint you are using.**

Latest, 7.3.0.

**The problem you want to solve.**

ESLint rules can have complex schemas. Some schemas can reach 100 lines long, and often allow various formats, such as in [no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports) which allows either an array of strings or an array of objects.

Today, unit tests for a rule will ideally ensure that the rule behavior is correct for all possible combinations of valid rule options, but it is not currently possible to test that a rule correctly disallows invalid rules schemas.

For example, with the rule [no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports), I would like to test that the rule schema fails validation when passed:
* Something that isn't an array
* An empty array
* An array containing an item that isn't a string nor object
* An array containing an object that is missing required properties like `name`
* Any other invalid combinations of input

Note that I am obviously not trying to test that JSON Schema works properly, but I am trying to test that I as a rule developer have written out my schema correctly. It can be tricky to get schemas to perfectly represent what the allowed input should be, especially when we don't even try to test the invalid cases.

The goal of improved testing here is to improve the quality of our rule schemas, thus reducing the chance of rule consumers configuring rules improperly (which can result in rules silently misbehaving or crashing).

**Your take on the correct solution to problem.**

In addition to the `valid` / `invalid` test case arrays in a rule unit test file, there could be an `error` array of invalid schemas.

**Are you willing to submit a pull request to implement this change?**

Likely yes.

**Comment by kaicataldo** [comment]
Since this would be a change to core, do you mind creating an [RFC](https://github.com/eslint/rfcs)? I agree that this would be a useful thing to be able to test for, and it would be especially nice for cases where one might want to use [custom keyword validators](https://ajv.js.org/#defining-custom-keywords).

Thoughts on figuring out a way to include these in the `invalid` cases array? I'm not sure there's a much value in adding a new `error` array, and I believe this could be done in a backwards-compatible fashion. As a quick illustration of what I'm envisioning, I _think_ something like the following might allow us to use the current API  (though changes would need to happen under the hood):

```js
var rule = require("../../../lib/rules/my-rule");
var RuleTester = require("eslint").RuleTester;

var ruleTester = new RuleTester();
ruleTester.run("my-rule", rule, {
    valid: ["bar", "baz"],
    invalid: [
        {
            code: "foo",
            options: ["some-option", { someOtherOption: true }]
            errors: [
                {
                    messageId: "invalidConfiguration":
                    data: {
                        message: "Expected configuration option someOtherOption to be a string. Found boolean instead."
                    }
                }
            ]
        }
    ]
});
```

**Comment by bmish** [comment]
@kaicataldo sure, I will work on putting an RFC together.

Using the `invalid` array sounds good. In the interest of keeping this lightweight and leaving out irrelevant properties, what do you think about:

1. Disallowing `code` from being passed when the config is being tested (since it can't affect the results).
2. Encouraging `messageId: "invalidConfiguration"` to be used by itself without the exact message string. I'm not trying to test JSON Schema and don't really care what message it outputs. In fact, we might even want to disallow testing the exact string so that future JSON Schema string changes won't break our test cases.

```js
{
    invalid: [
        {
            options: ["some-option", { someOtherOption: true }],
            errors: [
                {
                    messageId: "invalidConfiguration"
                }
            ]
        }
    ]
}
```

**Comment by kaicataldo** [comment]
Both your points make a lot of sense to me!

**Comment by bmish** [comment]
Reopening this as something I may try to write an RFC for at some point.

**Comment by bmish** [comment]
I have opened an RFC for this:
- https://github.com/eslint/rfcs/pull/103

**Comment by aladdin-add** [comment]
moved to "ready for implement", as the rfc has been merged. :)

**Comment by nzakas** [comment]
@eslint/eslint-team this has been ready to implement for two years. Does anyone want to take this on?

**Comment by snitin315** [comment]
I'll take this up
